ugvc/mrd/featuremap_consensus_utils.py

```diff
@@ -237,8 +251,64 @@ def pileup_featuremap(
             prev_key = rec_id
 
         # write last record
-        write_a_pileup_record(cons_dict[prev_key], prev_key, out_fh, header, min_qual)
+        if len(cons_dict.keys()) > 0:
+            write_a_pileup_record(cons_dict[prev_key], prev_key, out_fh, header, min_qual, sample_name, qual_agg_func)
 
     out_fh.close()
     pysam.tabix_index(output_vcf, preset="vcf", min_shift=0, force=True)
     return output_vcf
+
+
+def pileup_featuremap_on_an_interval_list(
+    featuremap: str,
+    output_vcf: str,
+    interval_list: str,
+    min_qual: int = 0,
+    sample_name: str = "SAMPLE",
+    qual_agg_func: callable = np.max,
+):
+    """
+    Apply pileup featuremap on an interval list
+
+    Inputs:
+        featuremap (str): The input featuremap vcf file
+        output_vcf (str): The output pileup vcf file
+        interval_list (str): The interval list file
+        min_qual (int): The minimum quality threshold
+        sample_name (str): The name of the sample (default: "SAMPLE")
+    Output:
+        output_vcf (str): The output vcf file
+    """
+    with tempfile.TemporaryDirectory(dir=dirname(output_vcf)) as temp_dir:
+        with open(interval_list, "r", encoding="utf-8") as f:
+            for line in f:
+                # ignore header lines
+                if line.startswith("@"):
+                    continue
+                # read genomic ineterval
+                genomic_interval = line.strip()
+                genomic_interval_list = genomic_interval.split("\t")
+                chrom = genomic_interval_list[0]
+                start = genomic_interval_list[1]
+                end = genomic_interval_list[2]
+                genomic_interval = chrom + ":" + str(start) + "-" + str(end)
+                # run pileup_featuremap on the interval
+                curr_output_vcf = pjoin(
+                    temp_dir,
+                    basename(output_vcf).replace(".vcf.gz", "")
+                    + "."
+                    + chrom
+                    + "_"
+                    + str(start)
+                    + "_"
+                    + str(end)
+                    + ".int_list.vcf.gz",
+                )
+                pileup_featuremap(featuremap, curr_output_vcf, genomic_interval, min_qual, sample_name, qual_agg_func)
+        # merge the output vcfs
+        vcfs = [pjoin(temp_dir, f) for f in os.listdir(temp_dir) if f.endswith(".vcf.gz")]
+        vcf_str = " ".join(vcfs)
+        cmd = f"bcftools concat {vcf_str} | bcftools sort - -Oz -o {output_vcf} && bcftools index {output_vcf}"
```

Comment:
1. bcftools index **-t**
2. Do you need to sort? I think an interval_list file is required to be sorted already - "The intervals MUST be sorted by coordinate (in increasing order) within contigs; and the contigs must be sorted in the same order as in the sequence dictionary." https://gatk.broadinstitute.org/hc/en-us/articles/360035531852-Intervals-and-interval-lists